File: app/src/main/java/com/elementary/tasks/core/view_models/notes/NoteViewModel.java
package com.elementary.tasks.core.view_models.notes;

import android.app.Application;

import com.elementary.tasks.core.data.models.Note;
import com.elementary.tasks.core.data.models.Reminder;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import static com.elementary.tasks.notes.preview.NotePreviewActivity.PREVIEW_IMAGES;

/**
 * Copyright 2018 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class NoteViewModel extends BaseNotesViewModel {

    public LiveData<Note> note;
    public LiveData<Note> editedPicture;
    public LiveData<Reminder> reminder;

    private NoteViewModel(Application application, String key) {
        super(application);
        note = getAppDb().notesDao().loadById(key);
        reminder = getAppDb().reminderDao().loadByNoteKey(key);
    }

    public void loadEditedPicture() {
        editedPicture = getAppDb().notesDao().loadById(PREVIEW_IMAGES);
    }

    public static class Factory extends ViewModelProvider.NewInstanceFactory {

        private Application application;
        private String key;

        public Factory(Application application, String key) {
            this.application = application;
            this.key = key;
        }

        @SuppressWarnings("unchecked")
        @Override
        public <T extends ViewModel> T create(Class<T> modelClass) {
            return (T) new NoteViewModel(application, key);
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/viewModels/notes/NoteViewModel.kt
package com.elementary.tasks.core.viewModels.notes

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.elementary.tasks.core.data.models.Note
import com.elementary.tasks.core.data.models.Reminder
import com.elementary.tasks.core.data.models.TmpNote
import com.elementary.tasks.core.utils.withUIContext
import com.elementary.tasks.core.viewModels.Commands
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.launch

/**
 * Copyright 2018 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class NoteViewModel private constructor(application: Application, key: String) : BaseNotesViewModel(application) {

    var note: LiveData<Note>
    val editedPicture: MutableLiveData<TmpNote> = MutableLiveData()
    var reminder: LiveData<Reminder>

    init {
        note = appDb.notesDao().loadById(key)
        reminder = appDb.reminderDao().loadByNoteKey(if (key == "") "1" else key)
    }

    fun saveTmpNote(byteArray: ByteArray?) {
        isInProgress.postValue(true)
        launch(CommonPool) {
            appDb.notesDao().insert(TmpNote(byteArray))
            withUIContext {
                isInProgress.postValue(false)
                result.postValue(Commands.IMAGE_SAVED)
            }
        }
    }

    fun deleteTmpNote(tmpNote: TmpNote) {
        isInProgress.postValue(true)
        launch(CommonPool) {
            appDb.notesDao().delete(tmpNote)
            withUIContext {
                isInProgress.postValue(false)
            }
        }
    }

    fun loadEditedPicture() {
        launch(CommonPool) {
            val note = appDb.notesDao().getEditedImage()
            withUIContext {  editedPicture.postValue(note) }
        }
    }

    class Factory(private val application: Application, private val key: String) : ViewModelProvider.NewInstanceFactory() {

        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return NoteViewModel(application, key) as T
        }
    }
}
